chiselFrontend/src/main/scala/chisel3/core/StrongEnum.scala

```diff
@@ -0,0 +1,339 @@
+
+  def all: List[T] = enum_instances.toList
+
+  def Value: T = {
+    val result = newEnum()
+    enum_instances.append(result)
+    enum_values.append(id)
+
+    width = (1 max id.bitLength).W
+    id += 1
+
+    // Instantiate all the enums when Value is called for the last time
+    if (enum_instances.length == enum_names.length && isTopLevelConstructor) {
+      bindAllEnums()
+    }
+
+    result
+  }
+
+  def Value(id: UInt): T = {
+    // TODO: These throw ExceptionInInitializerError which can be confusing to the user. Get rid of the error, and just
+    // throw an exception
+    if (!id.litOption.isDefined)
+      throw NonLiteralEnumException(s"$enumTypeName defined with a non-literal type in companion object")
+    if (id.litValue() <= this.id)
+      throw NonIncreasingEnumException(s"Enums must be strictly increasing: $enumTypeName")
+
+    this.id = id.litValue()
+    Value
+  }
+
+  def apply(): T = newEnum()
+
+  def apply(n: UInt)(implicit sourceInfo: SourceInfo, connectionCompileOptions: CompileOptions): T = {
+    if (!n.litOption.isDefined) {
+      throwException(s"Illegal cast from non-literal UInt to $enumTypeName. Use fromBits instead")
+    }
+
+    val result = enum_instances.find(_.litValue == n.litValue)
+
+    if (result.isEmpty) {
+      throwException(s"${n.litValue}.U is not a valid value for $enumTypeName")
+    } else {
+      result.get
+    }
+  }
+
+  def fromBits(n: UInt)(implicit sourceInfo: SourceInfo, connectionCompileOptions: CompileOptions): T = {
+    if (n.litOption.isDefined) {
+      apply(n)
+    } else if (!n.isWidthKnown) {
+      throwException(s"Non-literal UInts being cast to $enumTypeName must have a defined width")
+    } else if (n.getWidth > this.getWidth) {
+      throwException(s"The UInt being cast to $enumTypeName is wider than $enumTypeName's width ($getWidth)")
+    } else {
+      Builder.warning(s"A non-literal UInt is being cast to $enumTypeName. You can check that the value is legal by calling isValid")
+
+      val glue = Wire(new UnsafeEnum(width))
+      glue := n
+      val result = Wire(newEnum())
+      result := glue
+      result
+    }
+  }
+
+  // StrongEnum basically has a recursive constructor. It instantiates a copy of itself internally, so that it can
+  // make sure that all EnumType's inside of it were instantiated using the "Value" function. However, in order to
+  // instantiate its copy, as well as to instantiate new enums, it has to make sure that it has a no-args constructor
+  // as it won't know what parameters to add otherwise.
+
+  protected def checkEmptyConstructorExists(): Unit = {
+    try {
+      implicitly[ClassTag[T]].runtimeClass.getDeclaredConstructor()
+      getClass.getDeclaredConstructor()
+    } catch {
+      case ex: NoSuchMethodException => throw NoEmptyConstructorException(s"$enumTypeName does not have a no-args constructor. Did you declare it inside a class?")
+    }
+  }
+
+  private val isTopLevelConstructor: Boolean = {
```

Comment:
This is a clever way to do this, I'm guessing there isn't a way to do this with reflection?

Reply:
Not one that I can find, unfortunately. I also considered using global variables, but I'm honestly not sure whether that's more or less elegant.